### PR TITLE
Standardize mocha chai sinon setup - Closes #1216

### LINKED ISSUE #1216: Standardize mocha/chai/sinon setup
Child of #672 
E.g. sandbox

### LINKED ISSUE #1242: Wrong use of sinon.stub in unit test case
There are 9 occurrences of `sinon.stub` method in unit test cases due to which few test may pass false positive. 

#### Crrrent Usage: 
```
sinon.stub
```

#### Expected Usage: 
```
sinon.stub()
```
Note the missing parenthesis for stub method. Due to which the function are not actually stubbed instead these are assigned to stub generator. 

screenshot at jan 09 15-41-31

### PR DESCRIPTION
**Pull request by ManuGowda**
Closes #1216 
Closes #1242

Sandboxing `chai` `sinon` and `sinon sandbox` before running the test to avoid redundant module require to follow test suite standards.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -33,7 +33,15 @@
     "before": true,
     "beforeEach": true,
     "after": true,
-    "afterEach": true
+    "afterEach": true,
+    "chai": true,
+    "sinonChai": true,
```
Comment (willclarktech) [inline .eslintrc.json]: I don't think this needs to be exposed as a global.

File: test/functional/system/4.multisig.js [hunk 2]
```diff
@@ -13,11 +13,11 @@
  */
 'use strict';
 
-var chai = require('chai');
-var expect = require('chai').expect;
+
+
```
Comment (willclarktech) [inline test/functional/system/4.multisig.js]: Empty lines.

File: .eslintrc.json [hunk 3]
```diff
@@ -33,7 +33,15 @@
     "before": true,
     "beforeEach": true,
     "after": true,
-    "afterEach": true
+    "afterEach": true,
+    "chai": true,
```
Comment (willclarktech) [inline .eslintrc.json]: Does chai itself definitely need to be available globally?

File: test/mocha.opts [hunk 4]
```diff
@@ -1,2 +1,3 @@
 --timeout 1200s
 --exit
+--require ./test/setup.js
```
Comment (willclarktech) [inline test/mocha.opts]: Newline - do you have a plugin for editorconfig?

File: test/setup.js [hunk 5]
```diff
@@ -1,41 +1,33 @@
...
- * Removal or modification of this copyright notice is prohibited.
- */
 'use strict';
 
 require('colors');
+var chai = require('chai');
+var sinon = require('sinon');
+var sinonChai = require('sinon-chai');
 var supertest = require('supertest');
+process.env.NODE_ENV = 'test';
```
Comment (willclarktech) [inline test/setup.js]: Prefer a blank line after `require` statements.

File: test/common/helpers/api.js [hunk 6]
```diff
@@ -13,7 +13,7 @@
  */
 'use strict';
 
-var test = require('../../test');
+var test = testSuite;
```
Comment (willclarktech) [inline test/common/helpers/api.js]: This pattern loses the benefit of the more explicit name `testSuite` over `test`.

File: test/setup.js [hunk 7]
```diff
@@ -1,41 +1,33 @@
...
 
 // Root object
-var test = {};
+var testSuite = {};
```
Comment (willclarktech) [inline test/setup.js]: "Suite" to me sounds like a collection of tests. This is more like `testContext`/`testResources`/`testRun`/`testSetup` or something.

File: test/setup.js [hunk 8]
```diff
@@ -100,5 +92,12 @@ test._.mixin({
 	}
 }, { chain: false });
 
-// Exports
-module.exports = test;
+// See https://github.com/shouldjs/should.js/issues/41
```
Comment (willclarktech) [inline test/setup.js]: This project doesn't use should.js.

File: test/setup.js [hunk 9]
```diff
@@ -100,5 +92,12 @@ test._.mixin({
...
+Object.defineProperty(global, 'should', { value: chai.should() });
+global.chai = chai;
+global.expect = chai.expect;
+global.sinon = sinon;
+global.sinonChai = sinonChai;
+global.sandbox = sinon.sandbox.create();
```
Comment (willclarktech) [inline test/setup.js]: This doesn't match the documented way of creating sandboxes: http://sinonjs.org/releases/v3.2.1/sandbox/

File: test/setup.js [hunk 10]
```diff
@@ -1,41 +1,33 @@
...
 
 // Node configuration
-test.baseUrl = 'http://' + test.config.address + ':' + test.config.httpPort;
-test.api = supertest(test.baseUrl);
+testSuite.baseUrl = 'http://' + testSuite.config.address + ':' + testSuite.config.httpPort;
+testSuite.api = supertest(testSuite.baseUrl);
 
-test._ = require('lodash');
+// Custom lodash mixin
+testSuite._ = require('lodash');
```
Comment (willclarktech) [inline test/setup.js]: Is setting `_` as a global in its own right a possibility or too messy? There's a lot of boilerplate left just extracting it into local scope with `var _ = testSuite._`.

File: test/setup.js [hunk 11]
```diff
@@ -100,5 +92,12 @@ test._.mixin({
...
-module.exports = test;
+// See https://github.com/shouldjs/should.js/issues/41
+Object.defineProperty(global, 'should', { value: chai.should() });
+global.chai = chai;
+global.expect = chai.expect;
+global.sinon = sinon;
+global.sinonChai = sinonChai;
+global.sandbox = sinon.sandbox.create();
+global.supertest = supertest;
+global.testSuite = testSuite;
```
Comment (willclarktech) [inline test/setup.js]: Did you consider ALL_CAPS for this?

File: test/unit/modules/transactions.js [hunk 12]
```diff
@@ -13,15 +13,12 @@
  */
 'use strict';
 
-var chai = require('chai');
-var expect = require('chai').expect;
 var async = require('async');
-var sinon = require('sinon');
+var rewire = require('rewire');
```
Comment (willclarktech) [inline test/unit/modules/transactions.js]: 👌

File: test/integration/setup/shell.js [hunk 13]
```diff
@@ -44,6 +44,9 @@ module.exports = {
 		child.stdout.pipe(process.stdout);
 
 		child.on('close', function (code) {
+			console.log('==================================');
+			console.log(code);
+			console.log('==================================');
```
Comment (willclarktech) [inline test/integration/setup/shell.js]: Is this relevant to this PR?

File: test/unit/modules/transactions.js (hunk 12 above)
Comment (ManuGowda) [reply test/unit/modules/transactions.js]: I think this is because i removed the rewire below and clubbed all required together, so thats the reason.

File: .eslintrc.json (hunk 3 above)
Comment (ManuGowda) [reply .eslintrc.json]: Yes, chai is being used in `swaggerspec` `lisk_core` and `lisk_request_limit`

File: .eslintrc.json [hunk 14]
```diff
@@ -33,7 +33,15 @@
...
+    "sinonChai": true,
+    "expect": true,
+    "should": true,
```
[the remark this answers is not in the document]
Comment (ManuGowda) [reply .eslintrc.json]: yes both `should` and `expect` is being used across tests.

File: .eslintrc.json (hunk 3 above)
Comment (willclarktech) [reply .eslintrc.json]: 1. `test/common/swaggerSpec.js` uses Chai to extend it with a custom assertion. This could be moved to the setup (so that assertion could then be used in other tests if needed).
1. `test/unit/api/fittings/lisk_cors.js` and `lisk_request_limit` use Chai to set up `sinon-chai`, which is redundant given our new setup.

File: .eslintrc.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (willclarktech) [reply .eslintrc.json]: We should standardize this to only use one style as part of this issue.

File: test/common/ws/serverProcess.js [hunk 15]
```diff
@@ -16,9 +16,9 @@
 var randomstring = require('randomstring');
 var SocketCluster = require('socketcluster');
 var MasterWAMPServer = require('wamp-socket-cluster/MasterWAMPServer');
-var sinon = require('sinon');
 var WSClient = require('./client');
 var Promise = require('bluebird');
+var sinon = require('sinon');
```
Comment (willclarktech) [inline test/common/ws/serverProcess.js]: This shouldn't be necessary now.

File: test/functional/functional.js [hunk 16]
```diff
@@ -13,7 +13,7 @@
  */
 'use strict';
 
-var test = require('../test');
+var test = testContext;
```
Comment (willclarktech) [inline test/functional/functional.js]: This should be removed as well.

File: test/functional/http/get/accounts.multisignatures.js [hunk 17]
```diff
@@ -14,9 +14,8 @@
 'use strict';
 
 var test = require('../../functional.js');
-var _ = test._;
+
```
Comment (willclarktech) [inline test/functional/http/get/accounts.multisignatures.js]: When you're removing a line, it's best to remove it entirely rather than leaving lots of extra blank space. (Also applies in other locations.)

File: test/setup.js [hunk 18]
```diff
@@ -14,28 +14,35 @@
...
 require('colors');
+var chai = require('chai');
+var sinon = require('sinon');
+var sinonChai = require('sinon-chai');
 var supertest = require('supertest');
 
+process.env.NODE_ENV = 'test';
+
+chai.use(sinonChai);
 // Root object
```
Comment (willclarktech) [inline test/setup.js]: I'm not sure this comment is still helpful since the name is more explicit. What do you think?

File: test/setup.js [hunk 19]
```diff
@@ -14,28 +14,35 @@
...
-var test = {};
+var testContext = {};
 
-test.config = require('./data/config.json');
-test.config.root = process.cwd();
+testContext.config = require('./data/config.json');
+testContext.config.root = process.cwd();
 
 // Optional logging
```
Comment (willclarktech) [inline test/setup.js]: Likewise: this is pretty self-explanatory.

File: test/setup.js [hunk 20]
```diff
@@ -14,28 +14,35 @@
...
 }
 
 // Node configuration
-test.baseUrl = 'http://' + test.config.address + ':' + test.config.httpPort;
-test.api = supertest(test.baseUrl);
+testContext.baseUrl = 'http://' + testContext.config.address + ':' + testContext.config.httpPort;
+testContext.api = supertest(testContext.baseUrl);
 
-test._ = require('lodash');
+// Custom lodash mixin
```
Comment (willclarktech) [inline test/setup.js]: As above.

File: test/unit/modules/transactions.js (hunk 12 above)
Comment (willclarktech) [reply test/unit/modules/transactions.js]: Yes, nicer to have all required `node_modules` together.

File: test/common/ws/serverProcess.js (hunk 15 above)
Comment (ManuGowda) [reply test/common/ws/serverProcess.js]: The tests started to fail somehow even after making it global, so I added the sinon back and things started to work, weird behaviour.

File: test/setup.js (hunk 18 above)
Comment (ManuGowda) [reply test/setup.js]: yeah, I think that make sense, i will remove the comment, since the name is more explicit.

File: test/unit/api/ws/workers/peersUpdateRules.js [hunk 21]
```diff
@@ -45,7 +40,7 @@ describe('PeersUpdateRules', function () {
 		validConnectionId  = 'ABCDEF123456789';
 		validErrorCode = 4100;
 		peersUpdateRules = new PeersUpdateRules(slaveWAMPServerMock);
-		peersUpdateRules.slaveToMasterSender.send = sinon.stub(peersUpdateRules.slaveToMasterSender, 'send').callsArg(3, null);
+		peersUpdateRules.slaveToMasterSender.send = sinonSandbox.stub(peersUpdateRules.slaveToMasterSender, 'send').callsArg(3, null);
```
Comment (willclarktech) [inline test/unit/api/ws/workers/peersUpdateRules.js]: This assignment is an odd pattern. I don't think it does anything not covered by `sinonSandbox.stub(peersUpdateRules.slaveToMasterSender, 'send').callsArg(3, null);` on its own

File: test/unit/api/ws/workers/peersUpdateRules.js [hunk 22]
```diff
@@ -119,7 +114,7 @@ describe('PeersUpdateRules', function () {
 		it('should return an error from server when invoked with valid arguments and received error code', function (done) {
 			peersUpdateRules.slaveToMasterSender.send.restore();
```
Comment (willclarktech) [inline test/unit/api/ws/workers/peersUpdateRules.js]: These restores can be removed now that we have the global sandbox and `afterEach` hook.

File: test/unit/api/ws/workers/slaveToMasterSender.js [hunk 23]
```diff
@@ -32,15 +28,16 @@ describe('SlaveToMasterSender', function () {
 				options: {
 					authKey: 'valid auth key'
 				}
-			}
+			},
+			sendToMaster: 'sendToMaster'
```
Comment (willclarktech) [inline test/unit/api/ws/workers/slaveToMasterSender.js]: Prefer putting the stub in here directly via `sinonSandbox.stub()`.

File: test/unit/helpers/RPC.js [hunk 24]
```diff
@@ -116,7 +113,7 @@ describe('wsRPC', function () {
 		var validPort = 4000, validIp = '127.0.0.1';
 
 		beforeEach(function () {
-			initializeNewConnectionStub = sinon.stub(ClientRPCStub.prototype, 'initializeNewConnection');
+			initializeNewConnectionStub = sinonSandbox.stub(ClientRPCStub.prototype, 'initializeNewConnection');
 		});
 
 		afterEach(function () {
```
Comment (willclarktech) [inline test/unit/helpers/RPC.js]: `afterEach` hooks like this can be removed now we have the sandbox.

File: test/unit/logic/inTransfer.js [hunk 25]
```diff
@@ -403,62 +397,61 @@ describe('inTransfer', function () {
...
 			});
 
 			it('should call modules.accounts.mergeAccountAndGet with round = slots.calcRound result', function () {
-				expect(accountsStub.mergeAccountAndGet.calledWith(sinon.match({round: slots.calcRound(dummyBlock.height)}))).to.be.true;
+				expect(accountsStub.mergeAccountAndGet.calledWith(sinonSandbox.match({round: slots.calcRound(dummyBlock.height)}))).to.be.true;
 			});
+		});
 
-			describe('when modules.accounts.mergeAccountAndGet fails', function () {
+		describe('when modules.accounts.mergeAccountAndGet fails', function () {
```
Comment (willclarktech) [inline test/unit/logic/inTransfer.js]: Are these definitely supposed to be nested at this level?

File: test/unit/logic/inTransfer.js [hunk 26]
```diff
@@ -494,58 +487,58 @@ describe('inTransfer', function () {
...
-					});
+			it('should call callback with error', function () {
+				inTransfer.undo(trs, dummyBlock, sender, function (err) {
+					expect(err).not.to.be.empty;
 				});
 			});
+		});
 
-			describe('when modules.accounts.mergeAccountAndGet succeeds', function () {
+		describe('when modules.accounts.mergeAccountAndGet succeeds', function () {
```
Comment (willclarktech) [inline test/unit/logic/inTransfer.js]: As above.

File: test/unit/logic/outTransfer.js [hunk 27]
```diff
@@ -581,58 +572,58 @@ describe('outTransfer', function () {
...
 			});
 
 			it('should call modules.accounts.mergeAccountAndGet with round = slots.calcRound result', function () {
-				expect(accountsStub.mergeAccountAndGet.calledWith(sinon.match({round: slots.calcRound(dummyBlock.height)}))).to.be.true;
+				expect(accountsStub.mergeAccountAndGet.calledWith(sinonSandbox.match({round: slots.calcRound(dummyBlock.height)}))).to.be.true;
 			});
+		});
 
-			describe('when modules.accounts.mergeAccountAndGet fails', function () {
+		describe('when modules.accounts.mergeAccountAndGet fails', function () {
```
Comment (willclarktech) [inline test/unit/logic/outTransfer.js]: As above.

File: test/unit/logic/inTransfer.js (hunk 25 above)
Comment (ManuGowda) [reply test/unit/logic/inTransfer.js]: yes, the intransfer undo was nested incorrectly due to which the tests where failing.

File: .eslintrc.json [hunk 28]
```diff
@@ -25,15 +25,15 @@
...
   },
   "globals": {
-    "it": true,
-    "describe": true,
-    "before": true,
-    "beforeEach": true,
-    "after": true,
-    "afterEach": true
+    "expect": true,
+    "should": true,
```
[the remark this answers is not in the document]
Comment (ManuGowda) [reply .eslintrc.json]: This will be covered in #1295

File: test/common/swaggerSpec.js [hunk 29]
```diff
@@ -14,13 +14,9 @@
 'use strict';
 
 var chai = require('chai');
```
[the remark this answers is not in the document]
Comment (ManuGowda) [reply test/common/swaggerSpec.js]: since we had a discussion about not making chai as global variable, I have not made any changes for this, this particular file using chai to add some custom logic middleware, so I think we can keep this one as it is.

File: test/unit/api/ws/workers/slaveToMasterSender.js (hunk 23 above)
Comment (ManuGowda) [reply test/unit/api/ws/workers/slaveToMasterSender.js]: This issue will be fixed as part of #1307

File: test/unit/helpers/RPC.js (hunk 24 above)
Comment (willclarktech) [reply test/unit/helpers/RPC.js]: See #1307
